src/Aspire.Hosting/ApplicationModel/ExpressionResolver.cs

```diff
@@ -0,0 +1,113 @@
+                (EndpointProperty.Port, true) when HasBothHostAndPort() => await endpointReference.Property(EndpointProperty.TargetPort).GetValueAsync(cancellationToken).ConfigureAwait(false),
+                // If Container -> Exe, we need to go through the container host
+                (EndpointProperty.Host or EndpointProperty.IPV4Host, false) => endpointReference.ContainerHost,
+                (EndpointProperty.Url, _) => string.Format(CultureInfo.InvariantCulture, "{0}://{1}:{2}",
+                                                endpointReference.Scheme,
+                                                await EvalEndpointAsync(endpointReference, EndpointProperty.Host).ConfigureAwait(false),
+                                                await EvalEndpointAsync(endpointReference, EndpointProperty.Port).ConfigureAwait(false)),
+                _ => await endpointReference.Property(property).GetValueAsync(cancellationToken).ConfigureAwait(false)
+            };
+        }
+
+        async Task<string?> EvalExpressionAsync(ReferenceExpression expr)
+        {
+            // This logic is similar to ReferenceExpression.GetValueAsync, except that we recurse on
+            // our own resolver method
+            var args = new object?[expr.ValueProviders.Count];
+
+            for (var i = 0; i < expr.ValueProviders.Count; i++)
+            {
+                args[i] = await ResolveWithContainerSourceAsync(expr.ValueProviders[i], endpointUsage, preprocess, cancellationToken).ConfigureAwait(false);
+            }
+
+            return string.Format(CultureInfo.InvariantCulture, expr.Format, args);
+        }
+
+        return value switch
+        {
+            ConnectionStringReference cs => await ResolveWithContainerSourceAsync(cs.Resource.ConnectionStringExpression, endpointUsage, preprocess, cancellationToken).ConfigureAwait(false),
+            IResourceWithConnectionString cs => await ResolveWithContainerSourceAsync(cs.ConnectionStringExpression, endpointUsage, preprocess, cancellationToken).ConfigureAwait(false),
+            ReferenceExpression ex => await EvalExpressionAsync(ex).ConfigureAwait(false),
+            EndpointReference endpointReference => await EvalEndpointAsync(endpointReference, EndpointProperty.Url).ConfigureAwait(false),
+            EndpointReferenceExpression ep => await EvalEndpointAsync(ep.Endpoint, ep.Property).ConfigureAwait(false),
+            IValueProvider vp => await vp.GetValueAsync(cancellationToken).ConfigureAwait(false),
+            _ => throw new NotImplementedException()
+        };
+    }
+
+    internal static async ValueTask<string?> ResolveAsync(bool sourceIsContainer, IValueProvider valueProvider, CancellationToken cancellationToken)
+    {
+        // For each endpoint, store two bools, to track if the host and port are in use
+        var endpointUsage = new Dictionary<EndpointReference, bool[]>();
```

Comment:
Instead of an array this could be a tuple:

```suggestion
        var endpointUsage = new Dictionary<EndpointReference, (bool, bool)>();
```

Reply:
@davidfowl that was my first thoughts, but tuples are immutable, so didn't seem ideal here, since we need to update the values piecemeal.

Reply:
I just realized how painful it would be to do that yep. A small class would be a little better but this is fine.